src/bp/ui-shared/src/TreeView/index.tsx

```diff
@@ -0,0 +1,288 @@
+  postProcessing?: PostProcessing<T>
+  /** You can also parse nodes manually and provide them to the view */
+  nodes?: TreeNode<T>[]
+  /** Filters the displayed nodes (expands all folders when filtering) */
+  filter?: Filter
+  /** The full paths of nodes which should be expanded */
+  expandedPaths?: string[]
+  /** Ensure that elements having "field" set to "value" are displayed (their parent will be expanded) */
+  visibleElements?: { value: string; field: string }[]
+  /** Whether or not to highlight the folder's name on click */
+  highlightFolders?: boolean
+
+  onDoubleClick?: (element: T) => void
+  onClick?: (element: T) => void
+  onExpandToggle?: (node: TreeNode<T>, isExpanded: boolean) => void
+  onContextMenu?: (element: T) => JSX.Element | undefined
+}
+
+type TreeNode<T> = ITreeNode<T> & {
+  id: string
+  type: string
+  parent?: TreeNode<T>
+  fullPath: string
+  childNodes?: TreeNode<T>[]
+}
+
+type ElementRenderer<T> = (element: T) => { label: JSX.Element | string; icon?: any }
+type PostProcessing<T> = (nodes: TreeNode<T>[]) => TreeNode<T>[]
+
+interface Filter {
+  text: string
+  /** The field to compare text on the provided element */
+  field: string
+}
+
+interface BuildTreeParams<T> {
+  elements: T[]
+  filter?: Filter
+  nodeRenderer: ElementRenderer<T>
+  folderRenderer?: ElementRenderer<string>
+  postProcessing?: PostProcessing<T>
+  pathProps?: string
+}
+
+const TreeView = <T extends {}>(props: TreeViewProps<T>) => {
+  const [nodes, setNodes] = useState<TreeNode<T>[]>([])
+  const [expanded, setExpanded] = useState(props.expandedPaths || [])
+  const [, forceUpdate] = useReducer(x => x + 1, 0)
+
+  const { elements, filter, nodeRenderer, folderRenderer, postProcessing } = props
+
+  useEffect(() => {
+    if (!elements || (elements && !nodeRenderer)) {
+      return
+    }
+
+    const nodes = buildTree({ elements, nodeRenderer, filter, folderRenderer, postProcessing })
+    traverseTree(nodes, node => {
+      if (props.visibleElements?.find(x => node.nodeData?.[x.field] === x.value)) {
+        handleNodeExpansion(node.parent!, true)
+        node.parent!.isExpanded = true
+        node.isSelected = true
+      }
+
+      if (filter?.text) {
+        node.isExpanded = true
+      }
+    })
+
+    setNodes(nodes)
+  }, [elements, filter])
+
+  useEffect(() => {
+    props.nodes && setNodes(props.nodes)
+  }, [props.nodes])
+
+  useEffect(() => {
+    setExpanded(props.expandedPaths || [])
+  }, [props.expandedPaths])
+
+  const handleNodeExpansion = (node: TreeNode<T>, isExpanded: boolean) => {
+    isExpanded ? setExpanded([...expanded, node.fullPath]) : setExpanded(expanded.filter(x => x !== node.fullPath))
+    node.isExpanded = isExpanded
+
+    props.onExpandToggle?.(node, isExpanded)
+  }
+
+  const handleNodeClick = (selectedNode: TreeNode<T>) => {
+    if (selectedNode.nodeData) {
+      props.onClick?.(selectedNode.nodeData)
```

Comment:
`onClick?.(selectedNode.nodeData)` should there be a period there?

Reply:
Yup, optional chaining require a dot after the question mark before methods, array elements, etc